compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -112,6 +113,38 @@ impl MirPhase {
     }
 }
 
+/// Where a specific `mir::Body` comes from.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+#[derive(HashStable, TyEncodable, TyDecodable, TypeFoldable)]
+pub struct MirSource<'tcx> {
+    pub instance: InstanceDef<'tcx>,
+
+    /// If `Some`, this is a promoted rvalue within the parent function.
+    pub promoted: Option<Promoted>,
+}
+
+impl<'tcx> MirSource<'tcx> {
+    pub fn item(def_id: DefId) -> Self {
```

Comment:
```suggestion
    pub fn item(def: ty::WithOptParam<DefId>) -> Self {
```
my personal preference, makes it harder to get this wrong

Reply:
This isn't mine. I've just moved it from `transform/mod.rs`.

Reply:
yeah, I still think that this would be a good change though :thinking: you don't have to do this in this PR